File: c++/hand_sorter.cpp
#include <vector>
#include <algorithm>
#include <functional>
#include "hand.h"
#include "card.h"
#include "hand_identifier.h"
#include "hand_sorter.h"

using namespace std;

vector<function<bool(const Hand&, const Hand&)>> HandSorter::comparators = {
   compareHighCard, comparePair, compareTwoPair, compareThreeOfAKind, compareStraight, 
   compareFlush, compareFullHouse, compareFourOfAKind, compareStraightFlush, compareRoyalFlush
};

void HandSorter::sortHands(vector<Hand>& hands) {
    sortHandsByType(hands);
    sortTies(hands);
}

void HandSorter::sortHandsByType(vector<Hand>& hands) {
    sort(hands.begin(), hands.end(), greater<Hand>());
}

void HandSorter::sortTies(vector<Hand>& hands) {
    int startIndex = 0;
    int lastHandType = hands[0].getType();

    function<bool(const Hand&, const Hand&)> comparator;

    for (int i = 0; i < hands.size(); i++) {
        if (hands[i].getType() != lastHandType) {
            comparator = comparators[hands[startIndex].getType()];

            sort(hands.begin() + startIndex, hands.begin() + i, comparator);

            startIndex = i;
            lastHandType = hands[i].getType();
        }
        else if (i == hands.size() - 1) {
            comparator = comparators[hands[startIndex].getType()];

            sort(hands.begin() + startIndex, hands.end(), comparator);
        }
    }
}

//=============== Tie-Breaking Comparators ===============
//All return true is first hand is stronger
//and false if second hand is stronger

bool HandSorter::compareRoyalFlush(const Hand& h1, const Hand& h2) {
    vector<Card> cardList1 = h1.getSortedCards();
    vector<Card> cardList2 = h2.getSortedCards();

    return cardList1[0].getSuit() > cardList2[0].getSuit();
}

bool HandSorter::compareStraightFlush(const Hand& h1, const Hand& h2) {
    vector<Card> cardList1 = h1.getSortedCards();
    vector<Card> cardList2 = h2.getSortedCards();

    int highestCardComparison = compareHighestCard(cardList1, cardList2);
    if (highestCardComparison == 1)
        return false; 
    else if (highestCardComparison == -1)
        return true;
    else
        return cardList1[0].getSuit() > cardList2[0].getSuit();
}

bool HandSorter::compareFourOfAKind(const Hand& h1, const Hand& h2) {
    Card c1 = getCardOccuringNTimes(h1.getSortedCards(), 4)[0];
    Card c2 = getCardOccuringNTimes(h2.getSortedCards(), 4)[0];

    int value1 = c1.getValue();
    int value2 = c2.getValue();

    if (c1.getValue() == 0)
        value1 = 13;

    if (c2.getValue() == 0)
        value2 = 13;

    return value1 > value2;
}

bool HandSorter::compareFullHouse(const Hand& h1, const Hand& h2) {
    Card c1 = getCardOccuringNTimes(h1.getSortedCards(), 3)[0];
    Card c2 = getCardOccuringNTimes(h2.getSortedCards(), 3)[0];

    int value1 = c1.getValue();
    int value2 = c2.getValue();

    if (c1.getValue() == 0)
        value1 = 13;

    if (c2.getValue() == 0)
        value2 = 13;

    return value1 > value2;
}

bool HandSorter::compareFlush(const Hand& h1, const Hand& h2) {
    vector<Card> cardList1 = h1.getSortedCards();
    vector<Card> cardList2 = h2.getSortedCards();

    int highestCardComparison = compareHighestCard(cardList1, cardList2);
    if (highestCardComparison == 1)
        return false; 
    else if (highestCardComparison == -1)
        return true;
    else
        return cardList1[0].getSuit() > cardList2[0].getSuit();
}

bool HandSorter::compareStraight(const Hand& h1, const Hand& h2) {
    vector<Card> cardList1 = h1.getSortedCards();
    vector<Card> cardList2 = h2.getSortedCards();

    int highestCardComparison = compareHighestCard(cardList1, cardList2);
    if (highestCardComparison == 1)
        return false; 
    else if (highestCardComparison == -1)
        return true;
    else
        return cardList1.back().getSuit() > cardList2.back().getSuit();
}

bool HandSorter::compareThreeOfAKind(const Hand& h1, const Hand& h2) {
    Card c1 = getCardOccuringNTimes(h1.getSortedCards(), 3)[0];
    Card c2 = getCardOccuringNTimes(h2.getSortedCards(), 3)[0];

    int value1 = c1.getValue();
    int value2 = c2.getValue();

    if (c1.getValue() == 0)
        value1 = 13;

    if (c2.getValue() == 0)
        value2 = 13;

    return value1 > value2;
}

bool HandSorter::compareTwoPair(const Hand& h1, const Hand& h2) {
    vector<Card> pairs1 = getCardOccuringNTimes(h1.getSortedCards(), 2);
    vector<Card> pairs2 = getCardOccuringNTimes(h2.getSortedCards(), 2);

    int highestCardComparison = compareHighestCard(pairs1, pairs2);

    if (highestCardComparison == 1)
        return false;
    else if (highestCardComparison == -1)
        return true;

    vector<Card> kicker1 = getCardOccuringNTimes(h1.getSortedCards(), 1);
    vector<Card> kicker2 = getCardOccuringNTimes(h2.getSortedCards(), 1);

    int kickerCardComparison = compareHighestCard(kicker1, kicker2);

    if (kickerCardComparison == 1)
        return false;
    else if (kickerCardComparison == -1)
        return true;

    Card kickerCard1 = kicker1[0];
    Card kickerCard2 = kicker2[0];

    return kickerCard1.getSuit() > kickerCard2.getSuit();
}

bool HandSorter::comparePair(const Hand& h1, const Hand& h2) {
    vector<Card> pair1 = getCardOccuringNTimes(h1.getSortedCards(), 2);
    vector<Card> pair2 = getCardOccuringNTimes(h2.getSortedCards(), 2);

    int highestCardComparison = compareHighestCard(pair1, pair2);

    if (highestCardComparison == 1)
        return false;
    else if (highestCardComparison == -1)
        return true;

   vector<Card> singles1 = getCardOccuringNTimes(h1.getSortedCards(), 1);
   vector<Card> singles2 = getCardOccuringNTimes(h2.getSortedCards(), 1);

   int highestSingleComparison = compareHighestCard(singles1, singles2);

    if (highestSingleComparison == 1)
        return false;
    else if (highestSingleComparison == -1)
        return true;

    Card highestSingle1 = getHighestCard(singles1);
    Card highestSingle2 = getHighestCard(singles2);
    
    return highestSingle1.getSuit() > highestSingle2.getSuit();
}

bool HandSorter::compareHighCard(const Hand& h1, const Hand& h2) {
    int highestCardComparison = compareHighestCard(h1.getSortedCards(), h2.getSortedCards());

    if (highestCardComparison == 1)
        return false;
    else if (highestCardComparison == -1)
        return true;

    Card highestCard1 = getHighestCard(h1.getSortedCards());
    Card highestCard2 = getHighestCard(h2.getSortedCards());

    return highestCard1.getSuit() > highestCard2.getSuit();
}

//=============== Helpers ===============

//Returns:
// 1: If second list has higher card
//-1: If first list has higher card
// 0: If both lists have the same cards
int HandSorter::compareHighestCard(const vector<Card>& cardList1, const vector<Card>& cardList2) {
    vector<int> valueList1, valueList2;

    for (int i = 0; i < cardList1.size(); i++) {
        //Push value for first list
        if (cardList1[i].getValue() == 0) {
            if (HandIdentifier::isStraight(cardList1) && cardList1.size() == Hand::HAND_SIZE)
                valueList1.push_back(0);
            else
                valueList1.push_back(13);
        }
        else
            valueList1.push_back(cardList1[i].getValue());

        //Push value for second list
        if (cardList2[i].getValue() == 0) {
            if (HandIdentifier::isStraight(cardList2) && cardList1.size() == Hand::HAND_SIZE)
                valueList2.push_back(0);
            else
                valueList2.push_back(13);
        }
        else
            valueList2.push_back(cardList2[i].getValue());
    }

    sort(valueList1.begin(), valueList1.end(), greater<>());
    sort(valueList2.begin(), valueList2.end(), greater<>());

    for (int i = 0; i < valueList1.size(); i++) {
        if (valueList1[i] < valueList2[i])
            return 1;
        else if (valueList1[i] > valueList2[i])
            return -1;
    }

    return 0;
}

//Does not work for straights with aces as this always treats aces as high
Card HandSorter::getHighestCard(vector<Card> cardList) {
    sort(cardList.begin(), cardList.end());

    if (cardList[0].getValue() == 0)
        return cardList[0];

    return cardList.back();
}

//Expects: A sorted vector of cards
vector<Card> HandSorter::getCardOccuringNTimes(const vector<Card>& cardList, int n) {
    vector<Card> result;
    int cardOccurrence = 1;
    int lastCardValue = cardList[0].getValue();

    for (int i = 1; i < cardList.size(); i++) {
        if (cardList[i].getValue() != lastCardValue) {
            if (cardOccurrence == n)
                result.push_back(cardList[i-1]);

            cardOccurrence = 1;
            lastCardValue = cardList[i].getValue();
        }
        else
            cardOccurrence++;

        if (i == cardList.size() - 1) {
            if (cardOccurrence == n)
                result.push_back(cardList[i]);
        }
    }

    return result;
}

File: rust/src/hand_sorter.rs
use crate::card::Card;
use crate::hand;
use crate::hand::Hand;
use crate::hand_identifier;

const COMPARATORS: [fn(&Hand, &Hand) -> bool; 10] = [compare_royal_flush,compare_pair, compare_two_pair, compare_three_of_a_kind, compare_straight, compare_flush, compare_full_house, compare_four_of_a_kind, compare_straight_flush, compare_royal_flush];

pub fn sort_hands(hands: &mut Vec<Hand>) {
    sort_by_type(hands);
    sort_ties(hands);
}

fn sort_by_type(hands: &mut Vec<Hand>) {
    hands.sort();
    hands.reverse();
}

fn sort_ties(hands: &mut Vec<Hand>) {
    let mut start_index = 0;
    let mut last_type = hands[0].get_hand_type();
    let mut subarray_indices: Vec<(usize, usize)> = Vec::new();

    for (i, hand) in hands.iter().enumerate() {
        if hand.get_hand_type() != last_type {
            subarray_indices.push((start_index,i-1));

            start_index = i;
            last_type = hands[i].get_hand_type();
        }
        else if i == hands.len() - 1 {
            subarray_indices.push((start_index,i));
        }
    }

    for indices in subarray_indices {
        sort_subarray(&mut hands[indices.0..=indices.1]);
    }
}

fn sort_subarray(hands: &mut [Hand]) {
    let hand_type = hands[0].get_hand_type();

    let comparator = COMPARATORS[hand_type];

    hands.sort_by(|h1, h2| {
        if comparator(h1, h2) {
            std::cmp::Ordering::Greater
        } else {
            std::cmp::Ordering::Less
        }
    });

}

// =============== Comparators ===============
// All return:
// true if l1 < l2
// false if l1 > l2

fn compare_royal_flush(h1: &Hand, h2: &Hand) -> bool {
    let l1 = h1.get_sorted_cards();
    let l2 = h2.get_sorted_cards();

    l1[0].get_suit() < l2[0].get_suit()
}

fn compare_straight_flush(h1: &Hand, h2: &Hand) -> bool {
    let l1 = h1.get_sorted_cards();
    let l2 = h2.get_sorted_cards();

    let highest_card_comparison = compare_highest_card(&l1, &l2);

    if highest_card_comparison == 1 {
        return true;
    }
    else if highest_card_comparison == -1 {
        return false;
    }

    l1[0].get_suit() < l2[0].get_suit()
}

fn compare_four_of_a_kind(h1: &Hand, h2: &Hand) -> bool {
    let l1 = h1.get_sorted_cards();
    let l2 = h2.get_sorted_cards();

    let c1 = &get_cards_occuring_n_times(&l1, 4)[0];
    let c2 = &get_cards_occuring_n_times(&l2, 4)[0];

    let mut value1 = c1.get_value();
    let mut value2 = c2.get_value();

    if value1 == 0 {
        value1 = 13;
    }

    if value2 == 0 {
        value2 = 13;
    }

    value1 < value2
}

fn compare_full_house(h1: &Hand, h2: &Hand) -> bool {
    let l1 = h1.get_sorted_cards();
    let l2 = h2.get_sorted_cards();

    let c1 = &get_cards_occuring_n_times(&l1, 3)[0];
    let c2 = &get_cards_occuring_n_times(&l2, 3)[0];

    let mut value1 = c1.get_value();
    let mut value2 = c2.get_value();

    if value1 == 0 {
        value1 = 13;
    }

    if value2 == 0 {
        value2 = 13;
    }

    value1 < value2
}

fn compare_flush(h1: &Hand, h2: &Hand) -> bool {
    let l1 = h1.get_sorted_cards();
    let l2 = h2.get_sorted_cards();

    let highest_card_comparison = compare_highest_card(&l1, &l2);

    if highest_card_comparison == 1 {
        return true;
    }
    else if highest_card_comparison == -1 {
        return false;
    }

    l1[0].get_suit() < l2[0].get_suit()
}

fn compare_straight(h1: &Hand, h2: &Hand) -> bool {
    let l1 = h1.get_sorted_cards();
    let l2 = h2.get_sorted_cards();

    let highest_card_comparison = compare_highest_card(&l1, &l2);

    if highest_card_comparison == 1 {
        return true;
    }
    else if highest_card_comparison == -1 {
        return false;
    }

    let mut highest_card_suit1 = l1[l1.len() - 1].get_suit();
    let mut highest_card_suit2 = l2[l2.len() - 1].get_suit();

    if hand_identifier::is_royal_straight(&l1) {
        highest_card_suit1 = l1[0].get_suit();
    }

    if hand_identifier::is_royal_straight(&l2) {
        highest_card_suit2 = l2[0].get_suit();
    }

    highest_card_suit1 < highest_card_suit2
}

fn compare_three_of_a_kind(h1: &Hand, h2: &Hand) -> bool {
    let l1 = h1.get_sorted_cards();
    let l2 = h2.get_sorted_cards();

    let c1 = &get_cards_occuring_n_times(&l1, 3)[0];
    let c2 = &get_cards_occuring_n_times(&l2, 3)[0];

    let mut value1 = c1.get_value();
    let mut value2 = c2.get_value();

    if value1 == 0 {
        value1 = 13;
    }

    if value2 == 0 {
        value2 = 13;
    }

    value1 < value2
}

fn compare_two_pair(h1: &Hand, h2: &Hand) -> bool {
    let l1 = h1.get_sorted_cards();
    let l2 = h2.get_sorted_cards();

    let pairs1 = get_cards_occuring_n_times(&l1, 2);
    let pairs2 = get_cards_occuring_n_times(&l2, 2);

    let highest_card_comparison = compare_highest_card(&pairs1, &pairs2);

    if highest_card_comparison == 1 {
        return true;
    }
    else if highest_card_comparison == -1 {
        return false;
    }

    let kicker1 = get_cards_occuring_n_times(&l1, 1);
    let kicker2 = get_cards_occuring_n_times(&l2, 1);

    let kicker_card_comparison = compare_highest_card(&kicker1, &kicker2);

    if kicker_card_comparison == 1 {
        return true;
    }
    else if kicker_card_comparison == -1 {
        return false;
    }

    let kicker_card1 = &kicker1[0];
    let kicker_card2 = &kicker2[0];

    kicker_card1.get_suit() < kicker_card2.get_suit()
}


fn compare_pair(h1: &Hand, h2: &Hand) -> bool {
    let l1 = h1.get_sorted_cards();
    let l2 = h2.get_sorted_cards();

    let pair1 = get_cards_occuring_n_times(&l1, 2);
    let pair2 = get_cards_occuring_n_times(&l2, 2);

    let highest_card_comparison = compare_highest_card(&pair1, &pair2);

    if highest_card_comparison == 1 {
        return true;
    }
    else if highest_card_comparison == -1 {
        return false;
    }

    let singles1 = get_cards_occuring_n_times(&l1, 1);
    let singles2 = get_cards_occuring_n_times(&l2, 1);

    let highest_single_comparison = compare_highest_card(&singles1, &singles2);

    if highest_single_comparison == 1 {
        return true;
    }
    else if highest_single_comparison == -1 {
        return false;
    }

    let mut highest_single_suit1 = singles1[singles1.len() - 1].get_suit();
    let mut highest_single_suit2 = singles2[singles2.len() - 1].get_suit();

    if singles1[0].get_value() == 0 {
        highest_single_suit1 = singles1[0].get_suit();
    }

    if singles2[0].get_value() == 0 {
        highest_single_suit2 = singles2[0].get_suit();
    }

    highest_single_suit1 < highest_single_suit2
}

// =============== Helpers ===============

// Returns:
//  1 if l1 < l2
// -1 if l1 > l2
//  0 if l1 == l2
fn compare_highest_card(l1: &Vec<Card>, l2:&Vec<Card>) -> i32 {
    let mut value_list1: Vec<usize> = Vec::new();
    let mut value_list2: Vec<usize> = Vec::new();


    for i in 0..l1.len() {
        value_list1.push(l1[i].get_value());

        if l1[i].get_value() == 0 {
            value_list1[i] = 13;

            if l1.len() == hand::HAND_SIZE {
                if hand_identifier::is_straight(l1) && !hand_identifier::is_royal_straight(l1) {
                    value_list1[i] = 0;
                }
            }
        }

        value_list2.push(l2[i].get_value());

        if l2[i].get_value() == 0 {
            value_list2[i] = 13;

            if l2.len() == hand::HAND_SIZE {
                if hand_identifier::is_straight(l2) && !hand_identifier::is_royal_straight(l2) {
                    value_list2[i] = 0;
                }
            }
        }
    }

    value_list1.sort();
    value_list2.sort();

    value_list1.reverse();
    value_list2.reverse();

    for (value1, value2) in value_list1.iter().zip(value_list2.iter()) {
        if value1 < value2 {
            return 1;
        }
        else if value1 > value2 {
            return -1;
        }
    }

    0
}

pub fn get_cards_occuring_n_times(card_list: &Vec<Card>, n: usize) -> Vec<Card> {    
    let mut result: Vec<Card> = Vec::new();

    for card in card_list {
        if !result.contains(card) {
            if card_list.iter().filter(|&x| x == card).count() == n {
                result.push(card.clone())
            }
        }
    }

    result
}
